File: packages/javascript/test/os.test.ts
import assert from 'assert';
import { existsSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import { describe, it } from 'node:test';
import {
	runCommand,
	getArch,
	getBootTime,
	getCpu,
	getDiskSize,
	getDiskUsage,
	getFreeDiskSize,
	getCpuCount,
	getEndianness,
	getKernelVersion,
	getPlatform,
	getHomeDir,
	getHostname,
	getMachineId,
	getShell,
	getSid,
	getSystemUptime,
	getRamSize,
	getRamUsage,
	getFreeRamSize,
	getUsedRamSize,
	getTempDir,
	getUptime,
	getUsername
} from '../dist/node';
import { contains } from '../dist/verify';

const PLATFORM_NAMES = ['windows', 'macos', 'linux', 'freebsd', 'unknown'];

// Node's own vocabulary, which is the one the Python package normalises into.
const ARCHITECTURES = [
	'arm',
	'arm64',
	'ia32',
	'loong64',
	'mips',
	'mipsel',
	'ppc',
	'ppc64',
	'riscv64',
	's390',
	's390x',
	'x64'
];

describe('OS', () => {
	it('runCommand', async () => {
		assert.strictEqual(await runCommand('echo a'), 'a');
		assert.strictEqual(await runCommand('echo b'), 'b');
	});

	// `sort` reads standard input on every supported platform, so it never returns
	// while that pipe is open. The timeout is what fails the test: a regression here
	// hangs rather than throwing.
	it('runCommand does not wait on standard input', { timeout: 10000 }, async () => {
		assert.strictEqual(await runCommand('sort'), '');
	});

	it('getCpu', () => {
		const cpuName = getCpu();

		assert.strictEqual(cpuName.length > 0, true);
		assert.match(cpuName, /[a-zA-Z0-9]+/);
		assert.notEqual(cpuName, 'Unknown');
		// The model of the processor, not the architecture the process was built for.
		// Both packages answer with the same string, and an architecture name here
		// means the lookup fell through to a last resort.
		assert.doesNotMatch(cpuName, /^(arm|arm64|aarch64|x86|x86_64|amd64|i[3-6]86)$/i);
	});

	it('getHostname', async () => {
		const hostname = await getHostname();

		assert.strictEqual(hostname.length > 0, true);
		assert.match(hostname, /[a-zA-Z0-9]+/);
		assert.notEqual(hostname, 'Unknown');
		// A name may hold spaces of its own, but never around itself and never a
		// line ending left over from whatever was asked for it.
		assert.strictEqual(hostname, hostname.trim());
	});

	/*
	 * Sample Response:
	 * Windows: a642d9e1-6063-4da7-8ea8-2298f989d01d
	 * Linux: 5c6ee51d3e514eb4883e4373e320192c
	 * macOS: BAC04154-124A-56E1-BFEB-D6D94FE5DBC0
	 */
	it('getMachineId', async () => {
		const mId = await getMachineId();

		let regex;

		switch (process.platform) {
			case 'win32':
			case 'darwin':
			case 'freebsd':
				regex = /^[0-9a-zA-Z]{8}-[0-9a-zA-Z]{4}-[0-9a-zA-Z]{4}-[0-9a-zA-Z]{4}-[0-9a-zA-Z]{12}/;
				break;
			default:
				regex = /^[0-9a-zA-Z]{8}-?[0-9a-zA-Z]{4}-?[0-9a-zA-Z]{4}-?[0-9a-zA-Z]{4}-?[0-9a-zA-Z]{12}/;
				break;
		}

		assert.match(mId, regex);
		// The same machine answers with the same id, whichever call asks.
		assert.strictEqual(await getMachineId(), mId);
	});

	it('getRamSize', async () => {
		const ramSize = getRamSize();

		assert.strictEqual(ramSize.endsWith('GB'), true);
		assert.strictEqual(contains(ramSize, ['1', '2', '4', '7', '8', '14', '16', '32']), true);
	});

	// Example: S-1-5-21-406418252-5582013529-1321253100-2001
	it('getSid', async () => {
		const { platform } = process;

		if (platform !== 'win32' && platform !== 'darwin') {
			await assert.rejects(getSid(), /Not supported on this operating system/);
			return;
		}

		const sidResult = await getSid();

		assert.match(sidResult, /^S-1-[0-59]-\d{2}-\d{8,10}-\d{8,10}-\d{8,10}-[1-9]\d{1,9}/);
	});

	it('getUptime', () => {
		assert.strictEqual(typeof getUptime() === 'number', true);
		assert.strictEqual(typeof getUptime({ format: true }) === 'string', true);
		assert.strictEqual(getUptime({ floor: true }).toString().indexOf('.') === -1, true);
		assert.strictEqual((getUptime() as number) >= 0, true);
	});

	it('getFreeRamSize', () => {
		const size = getFreeRamSize();

		assert.match(size, /^\d+ [A-Z]+$/);
		// A running machine always has some memory free, so a zero here is a read
		// that failed rather than a machine that is full.
		assert.notEqual(size, '0 B');
	});

	it('getUsedRamSize', () => {
		assert.match(getUsedRamSize(), /^\d+ [A-Z]+$/);
	});

	it('getRamUsage', () => {
		const usage = getRamUsage();

		assert.strictEqual(typeof usage === 'number', true);
		assert.strictEqual(usage >= 0 && usage <= 100, true);
		assert.strictEqual(Number.isInteger(getRamUsage(0)), true);
		// The argument is how many decimal places are kept.
		assert.strictEqual((getRamUsage(3).toString().split('.')[1] || '').length <= 3, true);
	});

	it('getDiskSize', async () => {
		const size = await getDiskSize();

		assert.match(size, /^\d+ [A-Z]+$/);
		assert.notEqual(size, '0 B');
		// The current working directory is the default, named or not.
		assert.strictEqual(await getDiskSize(process.cwd()), size);
	});

	it('getFreeDiskSize', async () => {
		assert.match(await getFreeDiskSize(), /^\d+ [A-Z]+$/);
	});

	it('getDiskUsage', async () => {
		const usage = await getDiskUsage();

		assert.strictEqual(typeof usage === 'number', true);
		assert.strictEqual(usage >= 0 && usage <= 100, true);
		assert.strictEqual(Number.isInteger(await getDiskUsage(undefined, 0)), true);
	});

	it('the disk functions report a path that is not there', async () => {
		const missing = join(tmpdir(), 'qsu-no-such-directory-8f21');

		await assert.rejects(getDiskSize(missing));
		await assert.rejects(getFreeDiskSize(missing));
		await assert.rejects(getDiskUsage(missing));
	});

	it('getSystemUptime', () => {
		const seconds = getSystemUptime() as number;

		assert.strictEqual(typeof seconds === 'number', true);
		assert.strictEqual(seconds > 0, true);
		// The machine has been running at least as long as this process has.
		assert.strictEqual(seconds >= (getUptime() as number), true);
		assert.strictEqual(typeof getSystemUptime({ format: true }) === 'string', true);
		assert.strictEqual(getSystemUptime({ floor: true }).toString().indexOf('.') === -1, true);
	});

	it('getBootTime', () => {
		const bootTime = getBootTime();

		assert.strictEqual(bootTime instanceof Date, true);
		assert.strictEqual(bootTime.getTime() < Date.now(), true);
		// The boot time and the uptime are two readings of the same thing, so they
		// agree to within the moment it takes to read them twice.
		const fromUptime = Date.now() - (getSystemUptime() as number) * 1000;

		assert.strictEqual(Math.abs(bootTime.getTime() - fromUptime) < 2000, true);
	});

	it('getUsername', () => {
		const username = getUsername();

		assert.strictEqual(username.length > 0, true);
		assert.notEqual(username, 'Unknown');
		assert.strictEqual(username, username.trim());
	});

	it('getHomeDir', () => {
		const home = getHomeDir();

		assert.strictEqual(home.length > 0, true);
		assert.strictEqual(existsSync(home), true);
	});

	it('getTempDir', () => {
		const temp = getTempDir();

		assert.strictEqual(temp.length > 0, true);
		assert.strictEqual(existsSync(temp), true);
	});

	it('getShell', () => {
		const shell = getShell();

		assert.strictEqual(shell.length > 0, true);
		// A path to a program rather than a bare name, on either kind of system.
		assert.match(shell, process.platform === 'win32' ? /\\|\.exe$/i : /^\//);
	});

	it('getPlatform', () => {
		const name = getPlatform();

		assert.strictEqual(PLATFORM_NAMES.includes(name), true);
		// Every system these tests run on is one the table covers, so `unknown` here
		// means a name fell out of it.
		assert.notEqual(name, 'unknown');
		assert.strictEqual(name === 'windows', process.platform === 'win32');
		assert.strictEqual(name === 'macos', process.platform === 'darwin');
		assert.strictEqual(name === 'linux', process.platform === 'linux');
	});

	it('getArch', () => {
		const architecture = getArch();

		assert.strictEqual(ARCHITECTURES.includes(architecture), true);
		assert.strictEqual(architecture, process.arch);
	});

	it('getCpuCount', () => {
		const count = getCpuCount();

		assert.strictEqual(Number.isInteger(count), true);
		assert.strictEqual(count >= 1, true);
		// The same machine answers with the same number, whichever call asks.
		assert.strictEqual(getCpuCount(), count);
	});

	it('getKernelVersion', () => {
		const version = getKernelVersion();

		assert.strictEqual(version.length > 0, true);
		assert.notEqual(version, 'Unknown');
		assert.match(version, /\d/);
	});

	it('getEndianness', () => {
		assert.strictEqual(['BE', 'LE'].includes(getEndianness()), true);
	});
});

File: packages/dart/test/os_test.dart
import 'dart:io';

import 'package:qsu/qsu.dart';
import 'package:test/test.dart';

const List<String> _platformNames = <String>[
  'windows',
  'macos',
  'linux',
  'android',
  'ios',
  'unknown',
];

/// Node's own vocabulary, which is the one this package normalises into.
const List<String> _architectures = <String>[
  'arm',
  'arm64',
  'ia32',
  'riscv32',
  'riscv64',
  'x64',
];

final RegExp _size = RegExp(r'^\d+ [A-Z]+$');

void main() {
  group('OS', () {
    test('getPlatform', () {
      final String name = getPlatform();

      expect(_platformNames.contains(name), isTrue);
      // Every system these tests run on is one the table covers.
      expect(name, isNot('unknown'));
      expect(name == 'windows', Platform.isWindows);
      expect(name == 'macos', Platform.isMacOS);
      expect(name == 'linux', Platform.isLinux);
    });

    test('getArch', () {
      expect(_architectures.contains(getArch()), isTrue);
    });

    test('getKernelVersion', () {
      final String version = getKernelVersion();

      expect(version, isNotEmpty);
      expect(version, isNot('Unknown'));
      expect(version, matches(RegExp(r'\d')));

      // The kernel version, not the name the system is sold under. Windows is
      // where the two part company: `10.0.26100` against `11`.
      if (Platform.isWindows) {
        expect(version, matches(RegExp(r'^\d+\.\d+\.\d+')));
      }
    });

    test('getEndianness', () {
      expect(<String>['BE', 'LE'].contains(getEndianness()), isTrue);
    });

    test('getCpuCount', () {
      expect(getCpuCount(), greaterThanOrEqualTo(1));
    });

    test('getCpu', () {
      final String name = getCpu();

      expect(name, isNotEmpty);
      expect(name, isNot('Unknown'));
      // The model of the processor, not the architecture it was built for.
      expect(
          name,
          isNot(matches(RegExp(r'^(arm|arm64|aarch64|x86|x86_64|amd64)$',
              caseSensitive: false))));
    });

    test('getCpuSpeed', () {
      expect(getCpuSpeed(), greaterThanOrEqualTo(0));
    });

    test('getCpuUsage', () async {
      final num usage = await getCpuUsage(interval: 50);

      expect(usage, inInclusiveRange(0, 100));
      expect(await getCpuUsage(interval: 50, decimals: 0), isA<int>());
      // Nothing is sampled over no time at all, so there is nothing to report.
      expect(await getCpuUsage(interval: 0), 0);
    });

    test('getRamSize', () {
      expect(getRamSize(), matches(_size));
      expect(getRamSize(), isNot('0 B'));
    });

    test('getFreeRamSize', () {
      final String size = getFreeRamSize();

      expect(size, matches(_size));
      // A running machine always has some memory free.
      expect(size, isNot('0 B'));
    });

    test('getUsedRamSize', () {
      expect(getUsedRamSize(), matches(_size));
    });

    test('getRamUsage', () {
      expect(getRamUsage(), inInclusiveRange(0, 100));
      expect(getRamUsage(decimals: 0), isA<int>());
    });

    test('getDiskSize', () async {
      final String size = await getDiskSize();

      expect(size, matches(_size));
      expect(size, isNot('0 B'));
      // The current working directory is the default, named or not.
      expect(await getDiskSize(path: Directory.current.path), size);
    });

    test('getFreeDiskSize', () async {
      expect(await getFreeDiskSize(), matches(_size));
    });

    test('getDiskUsage', () async {
      final num usage = await getDiskUsage();

      // A disk with an operating system on it is never empty.
      expect(usage, greaterThan(0));
      expect(usage, lessThanOrEqualTo(100));
    });

    test('getUptime', () {
      expect(getUptime(), isA<num>());
      expect(getUptime() as num, greaterThan(0));
      expect(getUptime(format: true), isA<String>());
      expect((getUptime(floor: true) as num) % 1, 0);
    });

    test('getSystemUptime', () {
      final num seconds = getSystemUptime() as num;

      expect(seconds, greaterThan(0));
      // The machine has been running at least as long as this process has.
      expect(seconds, greaterThanOrEqualTo(getUptime() as num));
      expect(getSystemUptime(format: true), isA<String>());
    });

    test('getBootTime', () {
      final DateTime bootTime = getBootTime();

      expect(bootTime.isBefore(DateTime.now()), isTrue);
      // The boot time and the uptime are two readings of the same thing, so
      // they agree to within the moment it takes to read them twice.
      final DateTime fromUptime = DateTime.now().subtract(
          Duration(milliseconds: ((getSystemUptime() as num) * 1000).round()));

      expect(bootTime.difference(fromUptime).inSeconds.abs(), lessThan(2));
    });

    test('getProcessMemoryUsage', () {
      final String size = getProcessMemoryUsage();

      expect(size, matches(_size));
      expect(size, isNot('0 B'));
    });

    test('getHostname', () async {
      final String hostname = await getHostname();

      expect(hostname, isNotEmpty);
      expect(hostname, isNot('Unknown'));
      // A name may hold spaces of its own, but never around itself.
      expect(hostname, hostname.trim());
    });

    test('getOsName', () async {
      final String name = await getOsName();

      expect(name, isNotEmpty);

      if (Platform.isMacOS) {
        expect(name, startsWith('macOS'));
      } else if (Platform.isWindows) {
        expect(name, startsWith('Windows'));
      } else if (Platform.isLinux) {
        // The name the distribution gives itself, not the bare word `Linux`.
        expect(name, isNot('Linux'));
      }
    });

    test('getUsername', () {
      final String username = getUsername();

      expect(username, isNotEmpty);
      expect(username, isNot('Unknown'));
    });

    test('getHomeDir', () {
      expect(Directory(getHomeDir()).existsSync(), isTrue);
    });

    test('getTempDir', () {
      expect(Directory(getTempDir()).existsSync(), isTrue);
    });

    test('getShell', () {
      final String shell = getShell();

      expect(shell, isNotEmpty);
      // A path to a program rather than a bare name, on either kind of system.
      expect(shell,
          matches(Platform.isWindows ? RegExp(r'\\|\.exe$') : RegExp(r'^/')));
    });

    test('getLocalIp', () async {
      final String address = await getLocalIp();

      expect(address, matches(RegExp(r'^(\d{1,3}\.){3}\d{1,3}$')));
      expect(address, isNot('0.0.0.0'));
    });

    test('getMachineId', () async {
      final String id = await getMachineId();

      expect(
          id,
          matches(RegExp(r'^[0-9a-zA-Z]{8}-?[0-9a-zA-Z]{4}-?[0-9a-zA-Z]{4}'
              r'-?[0-9a-zA-Z]{4}-?[0-9a-zA-Z]{12}')));
      // The same machine answers with the same id, whichever call asks.
      expect(await getMachineId(), id);
    });

    test('getSid', () async {
      if (!Platform.isWindows && !Platform.isMacOS) {
        expect(getSid(), throwsUnsupportedError);

        return;
      }

      expect(await getSid(),
          matches(RegExp(r'^S-1-[0-59]-\d{2}-\d{8,10}-\d{8,10}')));
    });

    test('runCommand', () async {
      expect(await runCommand('echo a'), 'a');
      expect(await runCommand('echo b'), 'b');
    });
  });
}
